streams/src/main/java/org/apache/kafka/streams/state/ReadOnlyKeyValueStore.java

```diff
@@ -0,0 +1,52 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements.  See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.  You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.kafka.streams.state;
+
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+/**
+ * A key value store that only supports read operations
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+@InterfaceStability.Unstable
+public interface ReadOnlyKeyValueStore<K, V> {
+
+    /**
+     * Get the value corresponding to this key
+     *
+     * @param key The key to fetch
+     * @return The value or null if no value is found.
+     * @throws NullPointerException If null is used for key.
+     */
+    V get(K key);
+
+    /**
+     * Get an iterator over a given range of keys. This iterator MUST be closed after use.
+     *
+     * @param from The first key that could be in the range
+     * @param to The last key that could be in the range
+     * @return The iterator for this range.
+     * @throws NullPointerException If null is used for from or to.
+     */
+    KeyValueIterator<K, V> range(K from, K to);
```

Comment:
I don't think we can guarantee Ordering in this case as the range of keys will possibly be over multiple Stores, i.e, when you are querying you are actually using a Facade wrapping the underlying state stores - i don't think we want want to bring all the keys in to memory and sort them.

Taking snapshots could be very expensive - not sure we want to do this for every query. RocksDB does have built in support for this, though.

Handling updates/deletions etc has to be dealt with at the `StateStore` level, i.e, each store type will have different semantics. However, we probably should state that the returned iterator needs to be safe from Concurrent Modifications. Two examples.


**RocksDBStore**
```
store.put(1, "one");
store.put(2, "two");
store.put(3, "three");
store.put(4, "four");
final KeyValueIterator<Integer, String> range = store.range(1, 5);
store.delete(4);
store.put(2, "eight");
store.flush();
while (range.hasNext()) {
        System.out.println(range.next());
 }
```
Results:
```
KeyValue(1, one)
KeyValue(2, two)
KeyValue(3, three)
KeyValue(4, four)
```
**ConcurrentHashMap**
```
final ConcurrentHashMap<Integer, String> map = new ConcurrentHashMap<>();
map.put(1, "one");
map.put(2, "two");
map.put(3, "three");
map.put(4, "four");

final Iterator<Map.Entry<Integer, String>> iterator = map.entrySet().iterator();
map.remove(4);
map.put(2, "eight");
while (iterator.hasNext()) {
   System.out.println(iterator.next());
}
```
Results:
```
2=eight
1=one
3=three
```

I know - `ConcurrentHashMap` doesn't have a `range(..)` method, but it demonstrates that the iterator is safe. HashMap would throw a `ConcurrentModificationException`